### ISSUE TITLE
bug when simulation of gammapy datasets with an agnpy spectrum model

### ISSUE THREAD

**Issue by guillaumegrolleron** [cosimoNigro/agnpy#117]
When I have tried to simulate observations with a SpectrumDatasetOnOff.fake() with a Agnpy  Spectral Model defined as a wrapper of gammapy SpectralModel,  an error has occurred :

`ValueError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
operands could not be broadcast together with shapes (30,10) (300,) 
`

Thus, the problem appears when program tries to compute
`sed = (prefactor * epsilon * emissivity).to("erg cm-2 s-1")`
Indeed epsilon is a (30,10) array when emissivity is a 300 one. This sizes come from the fact that `self.geom.axes["energy_true"].edges is a (31,)` array.

I'm maybe wrong but I did not solve this error by modifiying the wrapper (cell 2 of https://agnpy.readthedocs.io/en/latest/tutorials/ssc_gammapy_fit.html).
I solved it by replacing the ligne 182 of the synchrotron module of agnpy:
`emissivity = integrator(integrand, gamma, axis=0)`
by :
`emissivity = integrator(integrand, gamma, axis=0).reshape(epsilon.shape)`

**Comment by cosimoNigro** [comment]
Thanks a lot for reporting this @guillaumegrolleron.

I am surprised that the `epsilon` array has dimension (30, 10)... 
It is converted from the `nu` that should be a one-dimensional array.
https://github.com/cosimoNigro/agnpy/blob/master/agnpy/synchrotron/synchrotron.py#L174

You encounter this problem only for the synchrotron SED?
This is the same problem you encountered in the fitting notebook, right?

**Comment by guillaumegrolleron** [comment]
Thanks for your answer @cosimoNigro 
In fact this bug is not specific for synchrotron because I have the same with inverse Compton.
This problem doesn't appear when I fit but when I try to simulate fake datasets, indeed when I use the `fake()` method, `npred_signal()` is called, which calls `compute_flux_spectral()` and then `integral()` and `integrate_spectrum()`. 
In this function the evaluator of the SpectralModel is computed with an energy defined as `energy = np.geomspace(energy_min, energy_max, num=int(num), axis=-1)` https://github.com/gammapy/gammapy/blob/ddbc5dc8a4291f125996a9ce0e9b18ecdcc4c6e5/gammapy/modeling/models/spectral.py#L97
`energy_min` and `energy_max` are the bounds of my `energy_axis` of size 30, and in this context num is equal to 10.
To sum up the problem is caused by the incompatibility between `integrate_spectrum()` (https://github.com/gammapy/gammapy/blob/ddbc5dc8a4291f125996a9ce0e9b18ecdcc4c6e5/gammapy/modeling/models/spectral.py#L78) which will call the evaluator of `Gammapy SpectralModel` on a multidimensional energy array and  the fact that the `evaluate_sed_flux()` method (https://github.com/cosimoNigro/agnpy/blob/86a28d49782d9f1546a4d21637bd0e175bb61ee1/agnpy/synchrotron/synchrotron.py#L123) is not made to be run with a frequency multidimensional array

**Comment by cosimoNigro** [comment]
Hello,

@guillaumegrolleron, first of all I am very sorry for such a slow reaction.
Finally I incorporated the sherpa and gammapy wrappers in the docs. Could you retry now and tell me if you have the same problem?

If that's the case, in the source code of the gammapy wrapper I left an inline comment
https://github.com/cosimoNigro/agnpy/blob/master/agnpy/fit/gammapy_wrapper.py#L178
pointing to this issue, with a tentative solution.

Let me know, and sorry again that it took a while.

**Comment by pivosb** [comment]
Just to complement the discussion:
Following this tutorial
https://agnpy.readthedocs.io/en/latest/tutorials/ssc_gammapy_mcmc_fit.html
I was trying to use instead of just FluxPointsDataset, a MapDataset (Fermi data) and a SpectrumDatasetOnOff (VERITAS data)
and I was getting the same error, and found this solution. However, I would get the same error in the SSC calculation, not only synchrotron. I solved by changing https://github.com/cosimoNigro/agnpy/blob/ee6c9667f3f0e1ef4ebd4978442a57684de59cf7/agnpy/synchrotron/synchrotron.py#L182 from

        emissivity = integrator(integrand, gamma, axis=0)

to

        emissivity = integrator(integrand, gamma, axis=0).reshape(epsilon.shape)

as suggested, but also in 
https://github.com/cosimoNigro/agnpy/blob/ee6c9667f3f0e1ef4ebd4978442a57684de59cf7/agnpy/compton/synchrotron_self_compton.py#L124 I changed from

        integral_epsilon = integrator(integral_gamma, epsilon, axis=0)
to

        integral_epsilon = integrator(integral_gamma, epsilon, axis=0).reshape(epsilon_s.shape)

But since I am not using SSA and/or EC, I am afraid that the same has to be done there, but I haven't checked.

**Comment by cosimoNigro** [comment]
Hello @pivosb, thanks for reporting this.

Are you using the latest version with the wrappers incorporated.
If so, did you try the solution I commented here in the code
https://github.com/cosimoNigro/agnpy/blob/master/agnpy/fit/gammapy_wrapper.py#L178

This problem never occurred to me while fitting flux points, it seems it is occurring either when using a `MapDataset` or doing a simulation, so I guess in those cases some array with energy and some other axis is passed to the functions evaluating the SED.

Would you mind checking if my solution at the wrapper level works?
Thank you.

**Comment by pivosb** [comment]
Hi @cosimoNigro.

Yes, I am. And yes, I first tried the solution suggested in 
https://github.com/cosimoNigro/agnpy/blob/master/agnpy/fit/gammapy_wrapper.py#L178

And you are correct, the problem is not with the flux points. It is specifically with any other dataset, and the error occurs before L178 in the wrapper. It happens when the code calls 

        sed_synch = Synchrotron.evaluate_sed_flux(
            nu, z, d_L, delta_D, B, R_b, self._n_e, *args, ssa=self.ssa
        )
        sed_ssc = SynchrotronSelfCompton.evaluate_sed_flux(
            nu, z, d_L, delta_D, B, R_b, self._n_e, *args, ssa=self.ssa
        )

and then the evaluate is called in [agnpy/agnpy/synchrotron/synchrotron.py](https://github.com/cosimoNigro/agnpy/blob/ee6c9667f3f0e1ef4ebd4978442a57684de59cf7/agnpy/synchrotron/synchrotron.py#L182) and [agnpy/agnpy/compton/synchrotron_self_compton.py](https://github.com/cosimoNigro/agnpy/blob/ee6c9667f3f0e1ef4ebd4978442a57684de59cf7/agnpy/compton/synchrotron_self_compton.py#L124).

So yes, it is related to using a `MapDataset` and/or a `SpectrumDatasetOnOff` (either a real one or a simulation)

**Comment by cosimoNigro** [comment]
Ok, thanks so it seems that my suggested fix will not help as the error occurs at the `evaluate_sed_flux` function calls. 

@pivosb would you mind fixing it in a PR?

**Comment by pivosb** [comment]
@cosimoNigro 

Yes, I can try! I will open a PR soon

**Comment by cosimoNigro** [comment]
Should be solved by @pivosb PR #131.
